src/java/picard/analysis/CollectGcBiasMetrics.java

```diff
@@ -83,187 +87,105 @@
     @Option(shortName = "BS", doc = "Whether the SAM or BAM file consists of bisulfite sequenced reads.  ")
     public boolean IS_BISULFITE_SEQUENCED = false;
 
-    // Used to keep track of the total clusters as this is kinda important for bias
-    private int totalClusters = 0;
-    private int totalAlignedReads = 0;
-    // Histograms to track the number of windows at each GC, and the number of read starts
-    // at windows of each GC. Need 101 to get from 0-100.
-    private static final int WINDOWS = 101;
-    private final int[] windowsByGc = new int[WINDOWS];
-    private final int[] readsByGc = new int[WINDOWS];
-    private final long[] basesByGc = new long[WINDOWS];
-    private final long[] errorsByGc = new long[WINDOWS];
-    private static int lastContig = SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX;
-    private byte[] gc;
-    private byte[] refBases;
-    private String saveHeader;
+    @Option(shortName = "LEVEL", doc = "The level(s) at which to accumulate metrics.  ")
+    private final Set<MetricAccumulationLevel> METRIC_ACCUMULATION_LEVEL = CollectionUtil.makeSet(MetricAccumulationLevel.ALL_READS);
 
-    /** Stock main method. */
+    // Calculates GcBiasMetrics for all METRIC_ACCUMULATION_LEVELs provided
+    private GcBiasMetricsCollector multiCollector;
+    public byte[] refBases;
+    public final int windowSize = WINDOW_SIZE;
+    final int[] windowsByGc = new int[WINDOWS];
+    public static final int WINDOWS = 101;
+    //Hash map of gc[] with reference name as key
+    private final Map<String, byte[]> gcByRef = new HashMap<String, byte[]>();
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Stock main method
+    ////////////////////////////////////////////////////////////////////////////
     public static void main(final String[] args) {
         System.exit(new CollectGcBiasMetrics().instanceMain(args));
     }
 
+    /////////////////////////////////////////////////////////////////////////////
+    // Setup calculates gc[] for the reference. Must be done at startup to avoid
+    // missing reference sequences in the case of small files that may
+    // not have reads aligning to every reference sequence
+    /////////////////////////////////////////////////////////////////////////////
     @Override
-    protected void setup(final SAMFileHeader header, final File samFile) {
+    protected void setup(final SAMFileHeader header, final File samFile, final File referenceFile) {
         IOUtil.assertFileIsWritable(CHART_OUTPUT);
         if (SUMMARY_OUTPUT != null) IOUtil.assertFileIsWritable(SUMMARY_OUTPUT);
-
-        try{
-            saveHeader = header.getReadGroups().get(0).getLibrary();
-        } catch (Exception ignored){
-            saveHeader = "LibraryNameUnavailable";
+        final ReferenceSequenceFile refFile = ReferenceSequenceFileFactory.getReferenceSequenceFile(REFERENCE_SEQUENCE);
```

Comment:
Is it possible for REFERENCE_SEQUENCE to be null?  Should handle this if so (i.e. in a CustomCommandLineValidation to prevent an ugly exception).